src/main/java/org/cascadebot/cascadebot/data/objects/GuildPermissions.java

```diff
@@ -5,9 +5,179 @@
+            result = Result.of(permission, PermissionAction.ALLOW, ResultCause.DISCORD, permission.getDiscordPerms());
+        }
+
+        return result;
+    }
+
+    private boolean hasDiscordPermissions(Member member, Channel channel, Set<Permission> permissions) {
+        if (CollectionUtils.isEmpty(permissions)) return false;
+        if (channel != null) {
+            return member.hasPermission(channel, permissions);
+        } else {
+            return member.hasPermission(permissions);
+        }
+    }
+
+    private Result evaluateMostRestrictiveMode(User user, List<Group> userGroups, CascadePermission permission) {
+        Result result = user.evaluatePermission(permission);
+        if (result.isDenied()) return result;
+
+        for (Group group : userGroups) {
+            Result groupResult = group.evaluatePermission(permission);
+            // If the result is neutral, it has no effect on the existing result.
+            if (groupResult.isNeutral()) continue;
+            // This is most restrictive mode so if any group permissions is DENY, the evaluated result is DENY.
+            if (groupResult.isDenied()) return groupResult;
+            result = groupResult;
+        }
+        return result;
+    }
+
+    private Result evaluateHierarchicalMode(User user, List<Group> userGroups, CascadePermission permission) {
+        Result result = Result.of(permission, PermissionAction.NEUTRAL);
+        // Loop through the groups backwards to preserve hierarchy; groups higher up override lower groups.
+        for (int i = userGroups.size() - 1; i >= 0; i--) {
+            Result groupResult = userGroups.get(i).evaluatePermission(permission);
+            if (groupResult.isNeutral()) continue;
+            // This overrides any previous action with no regard to what it was.
+            result = groupResult;
+        }
+
+        Result userResult = user.evaluatePermission(permission);
+        // User permissions take ultimate precedence over group permissions
+        if (!userResult.isNeutral()) {
+            result = userResult;
+        }
+
+        return result;
+    }
+
+    private Result getDefaultAction(CascadePermission permission) {
+        // A default permission will never explicitly deny a permission.
+        return permission.isDefaultPerm() ?
+                Result.of(permission, PermissionAction.ALLOW, ResultCause.DEFAULT) :
+                Result.of(permission, PermissionAction.DENY, ResultCause.DEFAULT);
+    }
+
+    public Group createGroup(String name) {
+        Set<String> ids = groups.stream().map(Group::getId).collect(Collectors.toSet());
+        Group group;
+        int iterations = 0;
+        do {
+            group = new Group(name);
+            if (++iterations == 7) {
+                // If this happens then... run?
+                CascadeBot.LOGGER.error("Somehow we couldn't manage to create a unique guild ID :(");
```

Comment:
"Somehow we couldn't manage to create a unique **guild** ID :("

might want to fix that lol